otherlibs/systhreads/thread.ml

```diff
@@ -19,6 +19,126 @@
 
 [@@@ocaml.flambda_o3]
 
+(* Implementation copied from [Domain.DLS], modulo thread-safety concerns. *)
+module TLS = struct
+  module Obj_opt : sig @@ portable
+    type t
+    val none : t
+    val some : 'a -> t
+    val is_some : t -> bool
+    val unsafe_get : t -> 'a
+  end = struct
+    type t = Obj.t
+    let none = Obj.magic_portable (Obj.repr (ref 0))
+    let some v = Obj.repr v
+    let is_some obj = (obj != Obj.magic_uncontended none)
+    let unsafe_get obj = Obj.obj obj
+  end
+
+  type tls_state = Obj_opt.t array
+
+  external get_tls_state
+    : unit -> tls_state @@ portable = "caml_thread_get_state"
+  external set_tls_state
+    : tls_state -> unit @@ portable = "caml_thread_set_state"
+
+  let init () =
+    let state = Array.make 8 (Obj.magic_uncontended Obj_opt.none) in
+    set_tls_state state
+
+  type 'a key = int * (unit -> 'a) Modes.Portable.t
+
+  let key_counter = Atomic.make 0
+
+  type key_initializer : value mod contended portable =
+      KI : 'a key * ('a -> (unit -> 'a) @ portable once) @@ portable
+      -> key_initializer
+  [@@unsafe_allow_any_mode_crossing "CR with-kinds"]
+
+  type key_initializer_list = key_initializer list
+
+  let parent_keys = Atomic.make ([] : key_initializer_list)
+
+  let rec add_parent_key ki =
+    let l = Atomic.get parent_keys in
+    if not (Atomic.compare_and_set parent_keys l (ki :: l))
+    then add_parent_key ki
+
+  let new_key ?split_from_parent init_orphan =
+    let idx = Atomic.fetch_and_add key_counter 1 in
+    let k = idx, { Modes.Portable.portable = init_orphan } in
+    begin match split_from_parent with
+    | None -> ()
+    | Some split -> add_parent_key (KI (k, split))
+    end;
+    k
+
+  (* If necessary, grow the current domain's local state array such that [idx]
+   * is a valid index in the array. *)
+  let maybe_grow idx =
+    let state = get_tls_state () in
+    let size = Array.length state in
+    if idx < size then state
+    else begin
+      let rec compute_new_size s =
+        if idx < s then s else compute_new_size (2 * s)
+      in
+      let new_size = compute_new_size size in
+      let new_state =
+        Array.make new_size (Obj.magic_uncontended Obj_opt.none)
+      in
+      Array.blit state 0 new_state 0 size;
+      set_tls_state new_state;
+      new_state
+    end
+
+  let set (type a) (idx, _init) (x : a) =
+    (* Assures [idx] is in range. *)
+    let state = maybe_grow idx in
+    (* [Sys.opaque_identity] ensures that flambda does not look at the type of
+     * [x], which may be a [float] and conclude that the [st] is a float array.
+     * We do not want OCaml's float array optimisation kicking in here. *)
+    Array.unsafe_set state idx (Obj_opt.some (Sys.opaque_identity x))
+
+  let get (type a) ((idx, init) : a key) : a =
+    (* Assures [idx] is in range. *)
+    let state = maybe_grow idx in
+    let obj = Array.unsafe_get state idx in
+    if Obj_opt.is_some obj
+    then (Obj_opt.unsafe_get obj : a)
+    else begin
```

Comment:
There is a similar opportunity here to use the inlined fast path &mdash; non-inlined slow path rewrite.